### PR TITLE
Backends overhaul

### LINKED ISSUE #666: Backend overhaul
## Description

Over time we've proven that Teleport does not need much of the complexity of the underlying storage back-ends. Examples of features that are not needed are:

* Global locks
* "Touching" values with TTL

Also, we have a new simpler file-based backend which:

* Provides a better key introspection and more compatible with traditional Linux admin tools.
* Provides a cheap and simple HA option if network-based filesystem is available or when Teleport is running inside a VM with live migration.

## Todo

* Remove global locks
* Switch to FS backed by default for new installs

### PR DESCRIPTION
**Pull request by kontsevoy**
This PR is a result of a massive code review and cleanup around Teleport backends.
There are two areas of focus:

* Usage of Global Locks
* DynamoDB backend

## DynamoDB Changes

- Added ability to read AWS config from `~/.aws` directory for testing
- Fixed TTL bug (UTC vs non-UTC comparison)
- Cleaned up buggy tests
- Proper implementation for `CreateVal()` (atomic one, with a proper "already exists" error)
- Renamed database attributes: the original implementation was using reserved DynamoDB keyword: "Key" [see here](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ReservedWords.html) for list of prohibited attribute names. **WARNING** this breaks existing DynamoDB installations, but this change HAD to be done.
- Added missing tests (for Create/Already Exists use case)
- Removed YAML->JSON->String configuration complexity and replaced it with simple pass-through 
- Re-implemented AcquireLock

## FS Backend

- FS backend now returns similar error types as Boltdb does
- Simpler AcquireLock implementation

## Other

I believe Teleport should eventually become lock-free. Meanwhile, the following two changes were made to move closer to that goal:

* Removed locks where they weren't critical. (i.e. "last writer wins" scenario is not bad).
* Added TTL restrictions to global lock: max of 1 minute.

## Breaking Changes

Existing DynamoDB clusters must be reconfigured. The original DynamoDB plugin was using "Key" as a "column name" which prevents DynamoDB from using this attribute in some queries.

Closes #666

### CODE REVIEW

**Comment by kontsevoy** [comment]
@klizhentas let me implement a clean (explicit) IAM support tomorrow (Tues). I'll ping you when this PR is ready.

**Comment by kontsevoy** [comment]
I have converted DynamoDB backend to use the default Go-SDK behavior which can be overridden via `teleport.yaml`, i.e.:

* First it looks in `~/.aws`
* Then it overrides with `AWS_xxx` environment variables
*  Then it overrides with `teleport.yaml`

... which is decent for now.

File: lib/auth/init.go [hunk 1]
```diff
@@ -105,12 +101,11 @@ func Init(cfg InitConfig, seedConfig bool) (*AuthServer, *Identity, error) {
...
 	}
 
-	lockService := local.NewLockService(cfg.Backend)
-	err := lockService.AcquireLock(cfg.DomainName, 60*time.Second)
+	err := cfg.Backend.AcquireLock(cfg.DomainName, 30*time.Second)
 	if err != nil {
 		return nil, nil, err
 	}
-	defer lockService.ReleaseLock(cfg.DomainName)
+	defer cfg.Backend.ReleaseLock(cfg.DomainName)
```
Comment (klizhentas) [inline lib/auth/init.go]: I'd prefer not to expose low-level key-value backend as an interface to auth service. That's why I had locking service. You can simply expose it as a part of `Presence` service to save typing.

File: lib/backend/backend.go [hunk 2]
```diff
@@ -78,3 +83,21 @@ type Config struct {
...
 	Tablename string `yaml:"table_name,omitempty"`
 }
+
+// Config is used for 'storage' config section. It's a combination of
+// values for various backends: 'boltdb', 'etcd', 'filesystem' and 'dynamodb'
+type Config struct {
+	// Type can be "bolt" or "etcd" or "dynamodb"
+	Type string `yaml:"type,omitempty"`
+
+	EtcdConfig   `yaml:",inline"`
```
Comment (klizhentas) [inline lib/backend/backend.go]: what if there are overlapping fields? There would be collision. I would like to avoid that.

**Review by klizhentas — CHANGES_REQUESTED** [review]
In addition to issues in the inline comments I'd like to address the fact that this is a breaking change for existing users. I think we should release a migration in the context of this PR to avoid the state with broken schemas.

File: lib/auth/init.go (hunk 1 above)
Comment (kontsevoy) [inline lib/auth/init.go]: Well... the goal is to remove locks completely. This PR reduces the number of lock-related code to an absolute minimum to discourage lock usage and simplify subsequent removal of them.

File: lib/backend/backend.go (hunk 2 above)
Comment (kontsevoy) [inline lib/backend/backend.go]: This has always been this way. This change does not affect runtim behavior. I have improved readability simply by grouping settings by a backend.

**Comment by kontsevoy** [comment]
Pushed requested changes... how do I mark the review as "responded to"?

**Review by klizhentas — APPROVED** [review]
approving assuming that #672 is on the way

**Pull request merged**